### PR TITLE
Added additional functional tests for ElementalValueProvider

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue - #1823.
Ported some tests from http://aspnetwebstack.codeplex.com/SourceControl/changeset/0e974218e12a67d4b1f25422119e40cfe6953e46#file_diff_test/System.Web.Http.Test/ModelBinding/ModelBindingEndToEndTests.cs
@pranavkm @yishaigalatzer

### CODE REVIEW

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs [hunk 1]
```diff
@@ -1692,5 +1693,215 @@ public class ModelBindingTest
...
+                { "ZipCode", "98052" },
+            };
+            var url = "http://localhost/BindModel/CollectionType";
+            var formData = new FormUrlEncodedContent(content);
+
+            // Act
+            var response = await client.PutAsync(url, formData);
+
+            // Assert
+            var address = JsonConvert.DeserializeObject<Address2>(
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs]: We should verify the status code before reading this. Can we move this to a separate method that does : 

```C#
public async Task<TVal> ReadValue<TVal>(HttpResponseMessage response)
{
   response.EnsureSuccessfulStatusCode();
   return JsonConvert.DeserializeObject<TVal>(await response.Content.ReadAsStringAsync());
}
```

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs [hunk 2]
```diff
@@ -1692,5 +1693,215 @@ public class ModelBindingTest
...
+            Assert.Null(person.Parent);
+        }
+
+        [Fact]
+        public async Task BindModelAsync_MultipleCheckBoxesWithSameKey_BindsFirstValue()
+        {
+            // Arrange
+            var server = TestServer.Create(_services, _app);
+            var client = server.CreateClient();
+            var content1 = new List<KeyValuePair<string, string>>
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs]: Make this a theory instead. They're two separate test cases.

File: test/WebSites/ModelBindingWebSite/Controllers/BindModelController.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNet.Mvc;
+using ModelBindingWebSite.Models;
+
+namespace ModelBindingWebSite.Controllers
+{
+    public class BindModelController : Controller
```
Comment (pranavkm) [inline test/WebSites/ModelBindingWebSite/Controllers/BindModelController.cs]: `Person_CollectionBinderController`

File: test/WebSites/ModelBindingWebSite/Models/Address2.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace ModelBindingWebSite.Models
+{
+    public class Address2
```
Comment (pranavkm) [inline test/WebSites/ModelBindingWebSite/Models/Address2.cs]: Needs a better name - `PersonAddress`?

File: test/WebSites/ModelBindingWebSite/Models/Address2.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace ModelBindingWebSite.Models
```
Comment (pranavkm) [inline test/WebSites/ModelBindingWebSite/Models/Address2.cs]: Could we move this to `ViewModels` namespace so we don't have two directories in the app with models?

File: test/WebSites/ModelBindingWebSite/Controllers/BindModelController.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        {
+            return user;
+        }
+
+        public PeopleModel NestedCollectionOfRecursiveTypes(PeopleModel model)
+        {
+            return model;
+        }
+
+        public bool PostCheckBox(bool isValid)
```
Comment (pranavkm) [inline test/WebSites/ModelBindingWebSite/Controllers/BindModelController.cs]: This doesn't necessarily hit the scenarios we'd want to test - leave this here as is - but could we get one where multiple values are selected:

```C#
public string CheckboxList(IEnumerable<UserPreferences> userPreferences)
{

}

public class UserPreference
{
   int Id { get; set; }
   
   bool Checked { get; set; }
}
```

And the post would look like `[0].Id = Foo&[0].Checked=true&[0].Checked=false&[1].Id=Bar&[1].Checked=false`

**Comment by pranavkm** [comment]
:watch:

**Comment by rynowak** [comment]
:shipit: from me. @pranavkm any more comments?

File: test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs [hunk 7]
```diff
@@ -1846,5 +1844,232 @@ public class ModelBindingTest
                     (actualErrorMessage) => actualErrorMessage.StartsWith(expectedErrorMessage));
             }
         }
+
+        public async Task BindModelAsync_WithCollection()
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingTest.cs]: `[Fact]`?

**Comment by pranavkm** [comment]
:shipit:

**Pull request merged**